client/remoterepo.c

```diff
@@ -200,12 +220,60 @@ TDNFDownloadFile(
         dwError = ERROR_TDNF_INVALID_PARAMETER;
         BAIL_ON_TDNF_ERROR(dwError);
     }
+    else
+    {
+        if(is_metalink)
+        {
+            if(fp)
+            {
+                fclose(fp);
+                fp = NULL;
+            }
+            fd = open(pszFile, O_RDONLY);
+            if(fd == -1)
+            {
+                dwError = errno;
+                BAIL_ON_TDNF_SYSTEM_ERROR(dwError);
+            }
+            while((length = read(fd, buf, sizeof(buf))) != 0)
+            {
+                r = metalink_parse_update(metalink_parser, buf, length);
+                if(r != 0)
+                {
+                    fprintf(stderr, "Unable to parse metalink, ERROR: code=%d\n", r);
+                    dwError = ERROR_TDNF_INVALID_PARAMETER;
+                    BAIL_ON_TDNF_ERROR(dwError);
+                }
+            }
+            r = metalink_parse_final(metalink_parser, NULL, 0, &metalink);
+            if((r != 0) || (metalink == NULL)) {
+                fprintf(stderr, "metalink_parse_final failed, ERROR: code=%d\n", r);
+                dwError = ERROR_TDNF_INVALID_PARAMETER;
+                BAIL_ON_TDNF_ERROR(dwError);
+            }
+            for(files = metalink->files; files && *files; ++files)
+            {
+                resources = (*files)->resources;
+                if(!resources) {
+                    continue;
+                }
+                dwError = TDNFRepoSetBaseUrl(pTdnf, pszRepo, (*resources)->url);
+                BAIL_ON_TDNF_ERROR(dwError);
+            }
+            /* delete metalink_t */
+            metalink_delete(metalink);
+        }
+    }     
 cleanup:
     TDNF_SAFE_FREE_MEMORY(pszUserPass);
     if(fp)
     {
         fclose(fp);
     }
+    if(fd > 0)
```

Comment:
check for >=0

Reply:
Priyesh, will open syacall ever return 0? I believe 0,1, 2 are for stdin, stout and stderr respectively. If we close fd 0,then we will be closing stdin right?

Reply:
open returns -1 on error. note that this code had set it to 0 as initial value. if the code above sets the correct initial value and operate on values per man, it makes for better code. concerns about closing 0 and such is not something this code need not worry about as it is clearly doing https://github.com/vmware/tdnf/pull/126/files/676df2608a8be16972cd7f8fd7e9258251d1b9a0#diff-a7c98e30fa73f96ee19509a659602e53R232 and such

Reply:
@sshedi, When fd is initialized to -1, then a check for `if (fd != -1) { close(fd); }` (in that case, a check for`fd > 0` would also work) is the best way to go about this. Any other construct, IMO, will confuse newcomers to the codebase, in spite of being *not* wrong. And yes, open will always return a number larger than 2 (unless someone closes FDs 0, 1 or 2 on purpose). 

@ppadmavilasom, I propose to drop the idea to check `fd >=0`; initialize `fd = -1`; and check for `fd != -1` to close. Thoughts?

Reply:
@cbsiddharth fd != -1 is good. i am okay with this

Reply:
@cbsiddharth @ppadmavilasom  Agree with this.

Reply:
I agree with first solution given by Priyesh, init fd to -1 and then check for (fd >= 0).
Reason why I say (fd != -1) is not so good is, we have to consider the possibility that in future open call may return other negative values than -1 and confused new comers will slowly come up to speed.

Reply:
@sshedi, its POSIX-ed (read as, written on stone) that open errors to -1. Also, fd of 0 being STDIN is the usual case, that not strictly true always (think of daemon processes). What if we closed FD 0 elsewhere (by design, hypothetically) and then when you call another open, then FD 0 will be be allocated to you **legally**. In such cases you are expected to call a close on FD 0 too.

Rule of thumb, if open didn't fail, the kernel allocated something for you by **whatever** FD it returned and you must call close on it. Even if FD == 0.

Reply:
I agree with the latter part of closing fd 0 and I mentioned it my previous comment
{(fd >=0) while closing}
But open errors to -1 and written on stone, I don't agree with it. It's software and can be changed anytime. TBH, something may replace POSIX entirely, like how POSIX replaced SystemV. So, nothing is written on stone in science.

Reply:
Well, I can argue against that :-). But, not without breaking most serious software in the word.

Reply:
Let's discuss further on this in person at office :)

Reply:
:)